pkg/controllers/externaloidc/externaloidc_controller.go

```diff
@@ -0,0 +1,373 @@
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/util/cert"
+	"k8s.io/utils/ptr"
+)
+
+const (
+	configNamespace           = "openshift-config"
+	managedNamespace          = "openshift-config-managed"
+	targetAuthConfigCMName    = "auth-config"
+	authConfigDataKey         = "auth-config.json"
+	oidcDiscoveryEndpointPath = "/.well-known/openid-configuration"
+)
+
+type externalOIDCController struct {
+	name                string
+	eventName           string
+	featureGateAccessor featuregates.FeatureGateAccess
+	authLister          configv1listers.AuthenticationLister
+	configMapLister     corev1listers.ConfigMapLister
+	configMaps          corev1client.ConfigMapsGetter
+}
+
+func NewExternalOIDCController(
+	featureGateAccessor featuregates.FeatureGateAccess,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	configInformer configinformers.SharedInformerFactory,
+	operatorClient v1helpers.OperatorClient,
+	configMaps corev1client.ConfigMapsGetter,
+	recorder events.Recorder,
+) factory.Controller {
+
+	c := &externalOIDCController{
+		name:      "ExternalOIDCController",
+		eventName: "external-oidc-controller",
+
+		featureGateAccessor: featureGateAccessor,
+		authLister:          configInformer.Config().V1().Authentications().Lister(),
+		configMapLister:     kubeInformersForNamespaces.ConfigMapLister(),
+		configMaps:          configMaps,
+	}
+
+	return factory.New().WithInformers(
+		// track openshift-config for changes to the provider's CA bundle
+		kubeInformersForNamespaces.InformersFor(configNamespace).Core().V1().ConfigMaps().Informer(),
+		// track auth resource
+		configInformer.Config().V1().Authentications().Informer(),
+	).WithFilteredEventsInformers(
+		// track openshift-config-managed/auth-config cm in case it gets changed externally
+		factory.NamesFilter(targetAuthConfigCMName),
+		kubeInformersForNamespaces.InformersFor(managedNamespace).Core().V1().ConfigMaps().Informer(),
+	).WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		ToController(c.name, recorder.WithComponentSuffix(c.eventName))
+}
+
+func (c *externalOIDCController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	if c.featureGateAccessor == nil || !c.featureGateAccessor.AreInitialFeatureGatesObserved() {
+		return nil
+	}
+
+	featureGates, err := c.featureGateAccessor.CurrentFeatureGates()
+	if err != nil {
+		return fmt.Errorf("could not get current feature gates: %v", err)
+	}
+
+	if !featureGates.Enabled(features.FeatureGateExternalOIDC) {
+		return nil
+	}
+
+	auth, err := c.authLister.Get("cluster")
+	if err != nil {
+		return fmt.Errorf("could not get authentication/cluster: %v", err)
+	}
+
+	if auth.Spec.Type != configv1.AuthenticationTypeOIDC {
+		// auth type is "IntegratedOAuth", "" or "None"; delete structured auth configmap if it exists
+		if err := c.configMaps.ConfigMaps(managedNamespace).Delete(ctx, targetAuthConfigCMName, metav1.DeleteOptions{}); err == nil {
```

Comment:
So at this line, we are in `!featureGates.Enabled(features.FeatureGateExternalOIDC) == false`, which means that the features is enabled. Now we check if it Auth is set to external OIDC and if it is, we attempt to delete the resource.

Wouldn't it be more beneficial to look up in the Informer Cache if it exists, before attempting to delete it? That way we could save a call to the apiserver? I mean this controller will be triggered a couple of times.